Sources/Harmony/Domain/Async/AsyncExecutor.swift

```diff
@@ -0,0 +1,69 @@
+//
+//  AsyncExecutor.swift
+//  Harmony
+//
+//  Created by Joan Martin on 10/6/22.
+//
+
+import Foundation
+
+/// Abstract definition of an async executor.
+@available(iOS 13.0.0, *)
+public protocol AsyncExecutor {
+    associatedtype T
+    func submit(operation: @escaping () async throws -> T) async throws -> T
+}
+
+/// Direct execution of the operation block.
+@available(iOS 13.0.0, *)
+public class AsyncDicrectExecutor<T>: AsyncExecutor {
+    public func submit(operation: @escaping () async throws -> T) async throws -> T {
+        try await operation()
+    }
+}
+
+
+/// Calls in the main thread (main actor) the operation block.
+@available(iOS 13.0.0, *)
+public class AsyncMainExecutor<T>: AsyncExecutor {
+    public func submit(operation: @escaping () async throws -> T) async throws -> T {
+        let task = Task { @MainActor in
+            return try await operation()
+        }
+        return try await task.value
+    }
+}
+
+/// Synchronizes a block execution of async calls.
+/// Concurrent calls of `submit` will result in serial exeution of their block code.
+@available(iOS 13.0.0, *)
+public actor AsyncAtomicExecutor<T>: AsyncExecutor  {
+    private enum Status {
+        case idle
+        case executing(Task<T,Error>)
+    }
+    private var status: Status = .idle
+    
+    public init() {}
+    
+    public func submit(operation: @escaping () async throws -> T) async throws -> T {
+        switch status {
+        case .idle:
+            let task: Task<T,Error> = Task {
+                let value = try await operation()
+                self.status = .idle
+                return value
+            }
+            status = .executing(task)
+            return try await task.value
+            
+        case .executing(let task):
+            // Waiting for the previous task to finish
+            _ = try? await task.value
+            // Checking for cancellation
+            try Task.checkCancellation()
+            // Re-submit the operation
+            return try await submit(operation: operation)
+        }
+    }
+}
```

Comment:
Please, correct me if I'm wrong.

With Harmony Executor, we were able to define a specific (serial or concurrent) thread to run on an executor. Giving us complete control over the resources and threads.

With Tasks, we do not have control over the threading. What we only control here is the queue of operations between different interactors. Every time an interactor is finished, the new interactor will use a new Task (with a new thread). We're just enforcing them to work on a serial execution. Is this enough?

I will ask again this question.
Why do we want to enforce the serial execution of interactors in the first place?  
* To reduce resources (having 1 Task at a time running, versus 100)

Reply:
**Why do we want to enforce the serial execution of interactors in the first place?**
As the paradigm changes here, I'd say no. We don't want this. We want to ensure our code can be run synchronously (awaiting stuff) and working with thread-safe implementations as much as we can.

The implementation of the AsyncAtomicExeuctor is just an exercise on how we could sincronize a block of code among multiple tasks in the new concurrent paradigm. To be used only when needed.

Using the latest Swift 5.7 version, we could agree to use in our interactions a generic `AsyncExecutor` which by default is a `AsyncDirectExeuctor` (which does nothing), and decide via DI if we need to protect some portions of our code (by using an `AsyncAtomicExecutor` or an `AsyncMainExecutor`) instead, similarly as we do now. Or maybe this is not even needed, and we are just too influenced by our old implementation?

**how can we run 2 operations concurrently with this implementation?**
This is the actual purpose of this class: to avoid this concurrency happening. This class pretends to force the code block to be executed serially, and never concurrently, even if you are being called from multiple concurrent tasks. Is mostly a "thread-safe" wall to hide portions of code that are sensible to concurrency.